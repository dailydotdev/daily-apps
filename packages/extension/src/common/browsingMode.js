/**
 * Lightweight script to detect whether the browser is running in Private mode.
 * @returns {Promise<boolean>}
 *
 */
export function isPrivateMode() {
    return new Promise(function detect(resolve) {
      var yes = function() { resolve(true); }; // is in private mode
      var not = function() { resolve(false); }; // not in private mode
  
      function detectChromeOpera() {
        // https://developers.google.com/web/updates/2017/08/estimating-available-storage-space
        var isChromeOpera = /(?=.*(opera|chrome)).*/i.test(navigator.userAgent) && navigator.storage && navigator.storage.estimate;
        if (isChromeOpera) {
          navigator.storage.estimate().then(function(data) {
            return data.quota < 120000000 ? yes() : not();
          });
        }
        return !!isChromeOpera;
      }
  
      function detectFirefox() {
        var isMozillaFirefox = 'MozAppearance' in document.documentElement.style;
        if (isMozillaFirefox) {
          if (indexedDB == null) yes();
          else {
            var db = indexedDB.open('inPrivate');
            db.onsuccess = not;
            db.onerror = yes;
          }
        }
        return isMozillaFirefox;
      }
  
      function detectEdgeIE10() {
        var isEdgeIE10 = !window.indexedDB && (window.PointerEvent || window.MSPointerEvent);
        if (isEdgeIE10) yes();
        return !!isEdgeIE10;
      }
  
      // when a browser is detected, it runs tests for that browser
      // and skips pointless testing for other browsers.
      if (detectChromeOpera()) return;
      if (detectFirefox()) return;
      if (detectSafari()) return;
      if (detectEdgeIE10()) return;
      
      // default navigation mode
      return not();
    });
  }
  