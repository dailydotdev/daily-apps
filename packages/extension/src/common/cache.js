export const STATE_KEY = 'state';
export const ANALYTICS_ID_KEY = 'analyticsId';
export const ANALYTICS_CONSENT_KEY = 'analyticsConsent';
export const TERMS_CONSENT_KEY = 'termsConsent';
export const DATA_VERSION_KEY = 'dataVersion';
export const CURRENT_DATA_VERSION = 2;

export function setCache(key, value) {
  return browser.storage.local.set({ [key]: value });
}

export async function getCache(key, defaultValue) {
  const res = await browser.storage.local.get(key);
  if (!(Number.isNaN(res[key]) || res[key] === undefined)) {
    return res[key];
  }

  return defaultValue;
}

export function removeCache(key) {
  return browser.storage.local.remove(key);
}
