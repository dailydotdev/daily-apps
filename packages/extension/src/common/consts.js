export const CRITICAL_FETCH_STAGE = 1;
export const OPERATIONAL_FETCH_STAGE = 2;
export const ENGAGEMENT_FETCH_STAGE = 4;
