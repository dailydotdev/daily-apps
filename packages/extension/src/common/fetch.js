// eslint-disable-next-line import/prefer-default-export
export const fetchTimeout = (url, ms, { signal, ...options } = {}) => {
  const controller = new AbortController();
  const promise = fetch(url, { signal: controller.signal, ...options });
  if (signal) {
    signal.addEventListener('abort', () => controller.abort());
  }
  const timeout = setTimeout(() => controller.abort(), ms);
  return promise.finally(() => clearTimeout(timeout));
};
