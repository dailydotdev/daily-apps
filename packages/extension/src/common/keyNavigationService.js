import {
  getPostByElement,
  getAbovePost,
  getBelowPost,
  getLeftPost,
  getRightPost,
  getTopLeftMostPostEl,
  hoverPost,
} from '@daily/components/src/common/domHelper';

export const validKeys = {
  h: 104,
  j: 106,
  k: 107,
  l: 108,
  '/': 47,
  b: 98,
};

function getNewPostEl(keyCode, currentElement, insaneMode) {
  if (keyCode === validKeys.h && !insaneMode) return getLeftPost(currentElement);

  if (keyCode === validKeys.l && !insaneMode) return getRightPost(currentElement);

  if (keyCode === validKeys.j) return getBelowPost(currentElement);

  if (keyCode === validKeys.k) return getAbovePost(currentElement);

  return currentElement;
}

function triggerBookmark(post) {
  if (post.ad) return null;

  return post.$emit('bookmark', { post: post.post, bookmarked: !post.post.bookmarked });
}

function getCurrentPost(posts, current) {
  if (!current) return null;

  const postOrAdProp = current.post || current.ad;

  return posts.find(article => [article.ad, article.post].indexOf(postOrAdProp) !== -1);
}

export function navigateDaily(keyCode, posts, enableSearch, { current, insaneMode }) {
  if (posts.length === 0) return null;

  const item = getCurrentPost(posts, current);

  if (Object.values(validKeys).indexOf(keyCode) === -1) return null;

  if (keyCode === validKeys['/']) return enableSearch();

  if (keyCode === validKeys.b && item) return triggerBookmark(item);

  const element = !item
    ? getTopLeftMostPostEl(posts, insaneMode)
    : getNewPostEl(keyCode, item.$el, insaneMode);

  const selectedPost = getPostByElement(posts, element);

  if (selectedPost === item) return selectedPost;

  hoverPost(selectedPost);

  return selectedPost;
}

Object.freeze(validKeys);

export default {
  navigateDaily,
};
