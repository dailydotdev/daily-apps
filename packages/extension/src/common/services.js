import {
  AuthServiceImpl, ContentServiceImpl, MonetizationServiceImpl, ProfileServiceImpl,
} from '@daily/services';

const app = 'extension';
// eslint-disable-next-line import/prefer-default-export
export const contentService = new ContentServiceImpl(process.env.VUE_APP_API_URL, 30, app);
export const monetizationService = new MonetizationServiceImpl(process.env.VUE_APP_API_URL, app);
export const profileService = new ProfileServiceImpl(process.env.VUE_APP_API_URL, app);
export const authService = new AuthServiceImpl(process.env.VUE_APP_API_URL, app);
