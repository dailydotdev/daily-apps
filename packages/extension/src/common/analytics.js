import { version } from './config';
import { browserName } from './browser';
import { getCache, ANALYTICS_ID_KEY } from './cache';

const queue = [];

const initialize = async () => {
  const clientId = await getCache(ANALYTICS_ID_KEY);
  if (window.location.protocol.indexOf('moz-extension') > -1) {
    let page;

    const getMessage = (action, type, ...args) => {
      if (action !== 'send') return null;

      const prefix = `v=1&tid=${process.env.VUE_APP_GA}&cid=${clientId}&aip=1&dp=${page}`;
      if (type === 'event') {
        return `${prefix}&t=event&ec=${encodeURIComponent(args[0])}&ea=${encodeURIComponent(args[1])}&el=${encodeURIComponent(args[2])}`;
      }
      if (type === 'pageview') {
        return `${prefix}&t=pageview`;
      }
      if (type === 'timing') {
        return `${prefix}&t=timing&utc=${encodeURIComponent(args[0])}&utv=${encodeURIComponent(args[1])}&utt=${encodeURIComponent(args[2])}`;
      }
      if (type === 'exception') {
        return `${prefix}&t=exception&exd=${encodeURIComponent(args[0].exDescription)}&exf=${args[0].exFatal ? 1 : 0}`;
      }

      return null;
    };

    window.ga = (action, type, ...args) => {
      if (action === 'set') {
        if (type === 'page') {
          page = encodeURIComponent(args[0]);
        }
        return;
      }

      const request = new XMLHttpRequest();
      const message = getMessage(action, type, ...args);
      if (message) {
        request.withCredentials = true;
        request.open('POST', 'https://www.google-analytics.com/collect', true);
        request.send(message);
      }
    };

    queue.forEach(args => window.ga(...args));
  } else {
    /* eslint-disable */
    (function (i, s, o, g, r, a, m) {
      a = s.createElement(o),
        m = s.getElementsByTagName(o)[0];
      a.async = 1;
      a.src = g;
      m.parentNode.insertBefore(a, m);
    })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
    /* eslint-enable */
    ga('create', process.env.VUE_APP_GA, { clientId });
    ga('set', 'checkProtocolTask', () => {
    });
    ga('require', 'displayfeatures');
  }
};

if (window.location.protocol.indexOf('moz-extension') > -1) {
  window.ga = (...args) => queue.push(args);
} else {
  /* eslint-disable */
  (function (i, s, o, g, r) {
    i['GoogleAnalyticsObject'] = r;
    i[r] = i[r] || function () {
      (i[r].q = i[r].q || []).push(arguments);
    }, i[r].l = 1 * new Date();
  })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
  /* eslint-enable */
}

export default function (consent) {
  if (consent) {
    return initialize();
  }

  window.ga = () => {
  };
  return Promise.resolve();
}

export const trackPageView = (page) => {
  const prefix = browserName ? '/extension' : '/';
  const suffix = browserName ? `v=${version}&b=${browserName}` : `v=${version}`;
  ga('set', 'page', `${prefix}${page}?${suffix}`);
  ga('send', 'pageview');
};
