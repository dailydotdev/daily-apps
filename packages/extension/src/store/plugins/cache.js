import { setCache, STATE_KEY } from '../../common/cache';
import { contentService } from '../../common/services';

const time2Cache = time => (time ? time.getTime() : null);

const vue2Json = x => ({ ...x });

const post2Cache = p => ({
  ...p,
  createdAt: time2Cache(p.createdAt),
  publishedAt: time2Cache(p.publishedAt),
});

const stateToCache = (state) => {
  // TODO: add tests
  const toCache = Object.assign({}, state);
  delete toCache.initialized;
  toCache.feed = {
    tags: state.feed.tags.filter(t => t.enabled).map(vue2Json),
    publications: state.feed.publications.map(vue2Json),
    posts: state.feed.posts.slice(0, contentService.pageSize).map(post2Cache),
    bookmarks: state.feed.bookmarks.map(post2Cache),
    latest: time2Cache(state.feed.latest),
  };
  const ui = Object.assign({}, state.ui);
  delete ui.showNotifications;
  delete ui.showDndMenu;
  delete ui.showSettings;
  toCache.ui = {
    ...ui,
    lastNotificationTime: time2Cache(state.ui.lastNotificationTime),
  };

  if (state.user.profile && state.user.profile.expiresIn) {
    toCache.user = {
      profile: {
        ...toCache.user.profile,
        expiresIn: time2Cache(state.user.profile.expiresIn),
      },
    };
  }

  return toCache;
};

const plugin = (store) => {
  store.subscribe((mutation, state) => {
    if (mutation.type !== 'loadFromCache' && mutation.type.indexOf('reset') < 0) {
      setCache(STATE_KEY, stateToCache(state))
      // eslint-disable-next-line no-console
        .catch(err => console.warn('failed to cache state', err));
    }
  });
};

export default plugin;
