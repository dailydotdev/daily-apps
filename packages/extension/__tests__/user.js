import module from '../src/store/modules/user';
import { authService, contentService, profileService } from '../src/common/services';
import { setCache, ANALYTICS_ID_KEY } from '../src/common/cache';
import { testAction } from './fixtures/helpers';

jest.mock('../src/common/services', () => ({
  authService: {
    authenticate: jest.fn(),
    logout: jest.fn(),
    getUserProfile: jest.fn(),
    generateChallenge: jest.fn(),
  },
  contentService: {
    setIsLoggedIn: jest.fn(),
  },
}));

jest.mock('../src/common/cache', () => ({
  ANALYTICS_ID_KEY: 'analytics',
  setCache: jest.fn(),
}));

beforeEach(() => {
  window.ga = () => {
  };
});

it('should set profile in state', () => {
  const state = { profile: null };
  const expected = { name: 'John', image: 'http://image.com' };
  module.mutations.setProfile(state, expected);
  expect(state.profile).toEqual(expected);
});

it('should set newUser to false in state', () => {
  const state = { profile: { newUser: true } };
  const expected = { profile: { newUser: false } };
  module.mutations.confirmNewUser(state);
  expect(state).toEqual(expected);
});

it('should authenticate user, set profile and check for bookmarks conflicts', async () => {
  const profile = {
    name: 'John',
    image: 'http://image.com',
  };
  authService.authenticate.mockReturnValue(profile);
  const state = { profile: null, challenge: { verifier: 'verifier' } };
  await testAction(
    module.actions.authenticate,
    { provider: 'google', code: '12345' },
    state,
    [{ type: 'feed/checkBookmarksConflicts', payload: null }, { type: 'setProfile', payload: profile }],
  );
  expect(authService.authenticate).toBeCalledWith('12345', 'verifier');
  expect(contentService.setIsLoggedIn).toBeCalledWith(true);
});

it('should authenticate user and set profile without checking for conflicts', async () => {
  const profile = {
    name: 'John',
    image: 'http://image.com',
    newUser: true,
  };
  authService.authenticate.mockReturnValue(profile);
  const state = { profile: null, challenge: { verifier: 'verifier' } };
  await testAction(
    module.actions.authenticate,
    { provider: 'google', code: '12345' },
    state,
    [{ type: 'setProfile', payload: profile }],
  );
  expect(authService.authenticate).toBeCalledWith('12345', 'verifier');
  expect(contentService.setIsLoggedIn).toBeCalledWith(true);
});

it('should do nothing when authentication fails', async () => {
  authService.authenticate.mockRejectedValue(null);
  const state = { profile: null };
  await testAction(
    module.actions.authenticate,
    { provider: 'google', code: '12345' },
    state,
    [],
  );
});

it('should logout and reset all preferences', async () => {
  authService.getUserProfile.mockReturnValue(Promise.resolve({ id: '1' }));
  const state = { profile: null };
  await testAction(
    module.actions.logout,
    null,
    state,
    [{ type: 'setProfile', payload: null }],
    [{ type: 'ui/reset', payload: null }, { type: 'feed/reset', payload: null }],
  );
  expect(setCache).toBeCalledWith(ANALYTICS_ID_KEY, '1');
  expect(contentService.setIsLoggedIn).toBeCalledWith(false);
});

it('should set code challenge in state', () => {
  const state = {};
  const expected = { verifier: 'verifier', challenge: 'challenge' };
  module.mutations.setChallenge(state, expected);
  expect(state.challenge).toEqual(expected);
});

it('should generate code challenge', async () => {
  authService.generateChallenge.mockReturnValue(Promise.resolve({
    verifier: 'verifier',
    challenge: 'challenge',
  }));
  const state = {};
  await testAction(
    module.actions.generateChallenge,
    null,
    state,
    [{ type: 'setChallenge', payload: { verifier: 'verifier', challenge: 'challenge' } }],
  );
  expect(authService.generateChallenge).toBeCalled();
});

it('should do nothing if not logged in', async () => {
  const state = {};
  await testAction(
    module.actions.validateAuth,
    null,
    state,
  );
});

it('should update profile if still logged in', async () => {
  const state = { profile: { id: '1', providers: ['github'] } };
  authService.getUserProfile.mockReturnValue(Promise.resolve({ id: '2', providers: ['github'] }));
  await testAction(
    module.actions.validateAuth,
    null,
    state,
    [{ type: 'setProfile', payload: { id: '2', providers: ['github'] } }],
  );
});

it('should logout if not logged in anymore', async () => {
  const state = { profile: { id: '1', providers: ['github'] } };
  authService.getUserProfile.mockReturnValue(Promise.resolve({ id: '2' }));
  await testAction(
    module.actions.validateAuth,
    null,
    state,
    [],
    [{ type: 'logout', payload: null }],
  );
});
