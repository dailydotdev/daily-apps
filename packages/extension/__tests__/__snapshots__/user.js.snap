// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should reset state 1`] = `
Object {
  "lastRead": null,
  "profile": null,
  "readingRank": Object {
    "nextRank": 0,
    "progress": 0,
    "rank": 0,
    "shownProgress": 0,
  },
}
`;
